### ISSUE TITLE
Restarted workers in kong pod use stale config

### ISSUE THREAD

**Issue by Benny-Git** [Kong/kong#9090]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

2.8.1

### Current Behavior

We switch a target URL on a service using the Admin API, and the change gets applied immediately. Everything is working as expected.
If however a worker is restarted (can be seen by message "start worker process" in logs), calls to the API suddenly arrive at the old target URL when routed via this worker.

### Expected Behavior

A newly started worker should always have the up-to-date config, and not rely on some stale cache.

### Steps To Reproduce

```markdown
I have been able to "reliably" recreate the issue in our DEV environment:

1. Creat service pointing to http://httpbin.utility:80/anything
2. Kill one one of the Kong DEV pods
3. On my laptop run watch -n 1 curl <route_url>
4. Requests go to upstream: http://192.168.99.186:80/anything with worker 1140
5. Edit service via to have path /get instead of /anything
6. Observe via logs the upstream changed to http://192.168.99.186:80/get with worker 1140
7. ps shows active workers 1140 and 1141
8. Remove one with kill 1140. New worker 1145 is created
9. Observe upstream is still http://192.168.99.186:80/get with worker 1141
10. Remove another one with kill 1141. New worker 1146 is created.
11. Observe upstream is now http://192.168.99.186:80/anything served by worker 1145 ⚡
12. After running curl -XDELETE localhost:8001/cache worker 1145 also serves http://192.168.99.186:80/get

Interestingly enough, repeating the steps above with switching the path from /get to /anything does not result in the issue -- this only worked reliably with switching the path from /anything to /get.
```

### Anything else?

We got hit by this several times now, even with instances where we created a new route that was then served, but after a worker restart did not get served anymore (receiving a "no Route matched with those values" instead).

We are running Kong on Kubernetes with 2 workers (default in Kong Helm chart), with Postgres as our database.

Are there any debug trace settings that could help pinpoint the root cause?

We originally reported this in https://github.com/Kong/kong/issues/8517

Benjamin Bertow benjamin.bertow@mercedes-benz.com, Mercedes-Benz Tech Innovation GmbH.
https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md

**Comment by Benny-Git** [comment]
> * From our limited trials, restarting the proxy pod does seem to resolve the problem.

A `curl -XDELETE localhost:8001/cache` should resolve the issue as well, without a restart.

**Comment by Benny-Git** [comment]
As a test we tried increasing the kong_mem_cache_size. The issue can still be recreated, unfortunately.

**Comment by Benny-Git** [comment]
I have recreated the issue from scratch with a simple setup now.

1. Start Kong locally using the [official Docker Compose template for Kong](https://github.com/Kong/docker-kong/tree/master/compose) (I used `make` to run it)
2. Create a simple test service
```json
POST localhost:8001/services/
{
    "name": "testservice",
    "url": "http://httpbin.org/anything"
}
```
3. Create a simple test route
```json
POST localhost:8001/services/testservice/routes
{
    "name": "testroute",
    "paths": ["/test"]
}
```
4. Start a loop querying the route
```bash
watch -n1 curl -s localhost:8000/test
``` 
The response will display the response coming from httpbin.org/anything
5. Dive into the running container with `docker exec -it kong-docker-compose_kong_1 bash`
6. List running processes with `ps a`
7. Kill nginx worker processes one by one using `kill <pid>` (for me there were 12 workers started)
8. Kong will immediately start a new worker when one is killed. However when you have gotten through the whole list from step 6, the loop from step 4 will suddenly display `{"message":"no Route matched with those values"}`
9. You can still list the routes and services, but they are not active in the workers anymore.
10. Clear cache with `DELETE localhost:8001/cache` and observe the loop coming back to life.
11. Repeat steps 6 and 7 to get the same result again. The only proper fix is a Kong restart (I used `make clean && make`), keeping the database intact

The recreate can be further simplified by setting `KONG_NGINX_WORKER_PROCESSES: 1` in docker-compose.yml, but I wanted to keep this as default as possible. The issue with stale cache also happens for other scenarios, e.g. changing the service url, or even deleting a route (it will reappear when the workers are killed).

**Comment by Benny-Git** [comment]
I have tried to analyze this using the standard Kong debug logs. Please find my debug log attached here: [kong_debug_logs.txt](https://github.com/Kong/kong/files/9159112/kong_debug_logs.txt)
The timeline:
```
11:40:39  --> started Kong with a DB where the service exists, but
 the route does not. Used 1 worker only. Initial worker PID 1110.
11:41:03  --> sent a curl to the non-existent route, resulting in a 404 (line 162)
11:41:13  --> created route (line 202)
11:42:46  --> killed worker 1110 (line 225)
          --> Master process 1 starts new worker 1262 (line 231)
11:43:34  --> sent another curl to the route, resulting a 404 (line 270)
```

From the debug log I don't see anything helpful. Is there any more fine-grained log level we can use to analyze this?

**Comment by ybce** [comment]
I was able to recreate the issue with the simple setup and it seems like this is also related to #8757

**Comment by Benny-Git** [comment]
@kikito @locao Did you have a chance to look into this?

Apart from creating unscheduled downtimes, I believe this could potentially also cause security issues
- if a deleted route makes an unexpected reappearance.
- if a consumer's password reverts to an older leaked instance
- if a newly created ip-restriction plugin gets removed again

Maybe someone more clever could come up with other scenarios I haven't foreseen yet.

**Comment by 23henne** [comment]
Any updates?

**Comment by locao** [comment]
Sorry for the long delay on this, we were overwhelmed with Kong 3.0. These [reproduction steps](https://github.com/Kong/kong/issues/9090#issuecomment-1190661854) are really helpful, thank you @Benny-Git. We are still looking for the culprit on this issue, the good news is that Kong 3.0 auto-recovers from it in a few seconds.

**Comment by bungle** [comment]
The PR to fix this is now merged. I think it can be backported to `2.8.x` series (but let's see).

**Comment by tingeltangelthomas** [comment]
@bungle we are struggling with this issue in Kong 2.8.x and we are not able to update to 3.x, now. 
How to ask officially for backporting the fix?

**Comment by Benny-Git** [comment]
It looks like this was not included in the 2.8.2 or 2.8.3 release?

**Comment by poelzi** [comment]
@Benny-Git  my pr  #9502 with a backport for 2.8.xwas just closed without a comment. kong, another project to avoid it seems....
